docs/lib/datastore/fragments/ios/auth-model.md

```diff
@@ -0,0 +1,170 @@
+        "plugins": {
+            "awsCognitoAuthPlugin": {
+                // ...
+            }
+        }
+    }
+```
+
+Add "AWSCognitoAuthPlugin" and "AWSAPIPlugin" to Amplify before configuring
+
+```swift
+Amplify.add(plugin: AWSCognitoAuthPlugin())
+Amplify.add(plugin: AWSAPIPlugin())
+Amplify.configure()
+```
+
+## Authentication requirements
+
+In order to issue an operation (`Datastore.save`, `Datastore.delete`) on a model that has fine grain controls defined, the user must be signed in using `Amplify.Auth.signIn`. A user's sign-in state may vary during the lifecycle of an app, so it is a common practice to check if the user has already been signed in by using `Amplify.Auth.fetchAuthSession` prior to attempting to update/delete a model that uses the @auth directive. The following are examples of how we recommend you handle auth state in your application:
+
+1. User's sign in state is known to be signed out, and the user needs to sign in, prior to making a call to `DataStore.save()`. 
+
+```swift
+Amplify.Auth.signIn {
+    switch $0 {
+    case .success(let signInResult):
+        if signInResult.isSignedIn {
+            // Amplify.DataStore.save
+        }
+    case .failure(let error):
+        print("Sign in error \(error)")
+    }
+}
+```
+
+2. User's sign in state is unknown (for example, app was terminated or transitioned from background to foreground) and you need to check prior to making a call to `Datastore.save()`. 
+
+```swift
+_ = Amplify.Auth.fetchAuthSession {
+    switch $0 {
+    case .success(let authSession):
+        if authSession.isSignedIn {
+            // Amplify.DataStore.save
+        }
+    case .failure(let error):
+        print("failed to get auth session", error)
+    }
+}
+```
+
+3. During the lifecycle of the app, listen to Auth events to show different views based on the user's state. See [Auth Events](~/lib/auth/auth-events.md) for more details.
+
+4. If the user state becomes signed out during the operation's sync to cloud, you can know exactly which operation failed when the error handler is called. By default, the error will be logged by `Amplify.Logging.error`
+
+```swift
+let configuration = DataStoreConfiguration.custom(errorHandler: { error in
+    Amplify.Logging.error(error: error)
+
+    if let dataStoreError = error as? DataStoreError,
+        case let .api(amplifyError, mutationEventOptional) = dataStoreError,
+        let actualAPIError = amplifyError as? APIError,
+        case let .operationError(_, _, underlyingError) = actualAPIError,
+        let authError = underlyingError as? AuthError,
+        let mutationEvent = mutationEventOptional {
+
+        if case .signedOut = authError {
+            let modelName = mutationEvent.modelName
+            let id = mutationEvent.modelId
+            let mutationType = mutationEvent.mutationType
+            print("The \(mutationType) operation for \(modelName) with \(id) failed due to user signed out.")
+        }
+    }
+})
+```
```

Comment:
I think this should be in a separate section - perhaps on its own page called "Configuration".

My thought is that in the first 3 examples, we are describing how the user should use datastore save/delete based on auth state.  In this case, I don't think we'd really recommend a customer to use a custom error handler to detect auth errors and then reactively call signin/save/delete --seems like they would just listen on the hub for the signin/expired/signed out events?

Reply:
one advantage is in the custom handler error handler, the developer can get the mutationEvent which failed. I'm not too sure what they would do with it other than, prompt the user to sign back in, and then maybe call DataStore.save with the mutation event data to retry. it's already such an advanced use case that i think for now let's remove it as you said